File: handlers/prospects.ts
import { UpnifyAuthenticator } from '../auth/upnifyAuth.js';
import { API_URLS, ENDPOINTS, DEFAULTS, UPNIFY_DEFAULTS } from '../utils/constants.js';
import { createErrorResponse, createSuccessResponse } from '../utils/validators.js';
import type { ProspectData } from '../types/interfaces.js';

export class ProspectsHandler {
    constructor(private auth: UpnifyAuthenticator) {}

    async createProspect(tkIntegracion: string, prospectData: ProspectData) {
        try {
            const { token, userInfo } = await this.auth.getTokenAndUserInfo(tkIntegracion);

            // Obtener el catálogo de fases para prospectos
            let tkFase = UPNIFY_DEFAULTS.TK_FASE;
            try {
                const phases = await this.auth.getProspectPhases(tkIntegracion);
                if (phases && phases.length > 0) {
                    tkFase = phases[0].tkFase;
                }
            } catch (phaseError) {
                // Usar fallback si no se pueden obtener las fases
                console.error('No se pudieron obtener las fases, usando fallback:', phaseError);
            }

            const upnifyPayload = {
                choice_empresa: "",
                search_terms: "",
                empresa: prospectData.empresa || "",
                tkEmpresa: "",
                cp: {
                    estatus: "",
                    validador: "",
                    division: "",
                    tipo: "",
                    gasto: "",
                    periodo: "",
                    tipoDeServicio: "",
                    testFecha: ""
                },
                nombre: prospectData.nombre || "",
                apellidos: prospectData.apellidos || "",
                titulo: "",
                sexo: prospectData.sexo || DEFAULTS.GENDER,
                correo: prospectData.correo || "",
                url: "",
                telefono2LadaPais: DEFAULTS.PHONE_PREFIX,
                telefono2: prospectData.telefono || "",
                movilLadaPais: DEFAULTS.PHONE_PREFIX,
                movil: prospectData.movil || "",
                puesto: prospectData.puesto || "",
                calle: prospectData.calle || "",
                colonia: prospectData.colonia || "",
                idPais: prospectData.idPais || DEFAULTS.COUNTRY,
                idEstado: "",
                idMunicipio: "",
                ciudad: prospectData.ciudad || "",
                codigoPostal: prospectData.codigoPostal || "",
                tkFase: tkFase,
                tkOrigen: "",
                facebook: "",
                twitter: "",
                skype: "",
                linkedIn: "",
                googlePlus: "",
                etiquetas: "",
                tkEtiquetas: ""
            };

            const response = await fetch(`${API_URLS.UPNIFY_BASE}${ENDPOINTS.PROSPECTS}`, {
                method: 'POST',
                headers: {
                    'token': token,
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify(upnifyPayload)
            });

            if (!response.ok) {
                const errorText = await response.text();
                console.error('Payload enviado:', JSON.stringify(upnifyPayload, null, 2));
                console.error('Token usado:', token);
                console.error('Respuesta del servidor:', errorText);
                throw new Error(`Error al crear prospecto: ${response.status} ${response.statusText}. ${errorText}`);
            }

            const result = await response.text();
            return {
                success: true,
                message: 'Prospecto creado exitosamente',
                response: result,
                tkEmpresa: userInfo.tkEmpresa
            };
        } catch (error) {
            throw new Error(`Error al crear prospecto en Upnify: ${error instanceof Error ? error.message : error}`);
        }
    }

    async searchContacts(tkIntegracion: string, searchParams: { buscar: string; cantidadRegistros: number }) {
        try {
            const { token, userInfo } = await this.auth.getTokenAndUserInfo(tkIntegracion);

            const queryParams = new URLSearchParams({
                buscar: searchParams.buscar,
                cantidadRegistros: searchParams.cantidadRegistros.toString(),
                secciones: DEFAULTS.SEARCH_SECTIONS
            });

            const response = await fetch(`${API_URLS.UPNIFY_BASE}${ENDPOINTS.SEARCH}?${queryParams}`, {
                method: 'GET',
                headers: {
                    'token': token,
                    'Content-Type': 'application/json',
                    'User-Agent': 'UpnifyMCP/1.0'
                }
            });

            if (!response.ok) {
                const errorText = await response.text();
                console.error('Parámetros enviados:', searchParams);
                console.error('Token usado:', token);
                console.error('Respuesta del servidor:', errorText);
                throw new Error(`Error al buscar contactos: ${response.status} ${response.statusText}. ${errorText}`);
            }

            const resultados = await response.json();

            if (!Array.isArray(resultados) || resultados.length < 2) {
                return {
                    success: true,
                    message: 'Búsqueda completada sin resultados',
                    termino: searchParams.buscar,
                    total: 0,
                    resumen: { prospectos: 0, clientes: 0 },
                    contactos: [],
                    tkEmpresa: userInfo.tkEmpresa
                };
            }

            // resultados[0] contiene el resumen, resultados[1] contiene los datos
            const resumen = resultados[0][0] || {};
            const contactosRaw = resultados[1] || [];

            // Filtrar y formatear solo prospectos y clientes
            const contactos = contactosRaw
                .filter((item: any) => item.seccion === 'prospectos' || item.seccion === 'clientes')
                .map((contacto: any) => ({
                    seccion: contacto.seccion,
                    tkProspecto: contacto.tkProspecto,
                    contacto: contacto.contacto,
                    correo: contacto.correo,
                    telefono: contacto.telefono,
                    movil: contacto.movil,
                    ejecutivo: contacto.ejecutivo,
                    ejecutivoIniciales: contacto.ejecutivoIniciales,
                    empresa: contacto.empresa || '',
                    id: `${contacto.seccion}-${contacto.tkProspecto}`
                }));

            return {
                success: true,
                message: 'Búsqueda completada exitosamente',
                termino: searchParams.buscar,
                total: contactos.length,
                resumen: {
                    prospectos: resumen.prospectos || 0,
                    clientes: resumen.clientes || 0
                },
                contactos: contactos,
                mensaje: contactos.length > 1
                    ? `Se encontraron ${contactos.length} contactos. Para crear una oportunidad, especifica el tkProspecto del contacto deseado.`
                    : contactos.length === 1
                        ? 'Se encontró 1 contacto exacto'
                        : 'No se encontraron contactos',
                tkEmpresa: userInfo.tkEmpresa
            };

        } catch (error) {
            throw new Error(`Error al buscar contactos en Upnify: ${error instanceof Error ? error.message : error}`);
        }
    }
}

File: python-package/src/upnify_mcp_server/prospects.py
"""
Prospects handler for Upnify MCP Server
"""

import httpx
from typing import Dict, Any, Optional
from .constants import API_URLS, ENDPOINTS, DEFAULTS
from .auth import UpnifyAuthenticator


class ProspectsHandler:
    """Handles prospects-related operations"""
    
    def __init__(self, authenticator: UpnifyAuthenticator):
        self.auth = authenticator
    
    async def create_prospect(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new prospect in Upnify"""
        auth_info = await self.auth.get_token_and_user_info()
        token = auth_info["token"]
        
        # Validate required fields
        if not data.get("nombre") or not data.get("correo"):
            raise ValueError("Los campos 'nombre' y 'correo' son obligatorios")
        
        # Prepare prospect data
        prospect_data = {
            "nombre": data["nombre"],
            "apellidos": data.get("apellidos", ""),
            "correo": data["correo"],
            "telefono": data.get("telefono", ""),
            "movil": data.get("movil", ""),
            "sexo": data.get("sexo", "H"),
            "puesto": data.get("puesto", ""),
            "empresa": data.get("empresa", ""),
            "ciudad": data.get("ciudad", ""),
            "idPais": data.get("idPais", DEFAULTS["DEFAULT_COUNTRY"]),
            "calle": data.get("calle", ""),
            "colonia": data.get("colonia", ""),
            "codigoPostal": data.get("codigoPostal", ""),
            "comentarios": data.get("comentarios", "")
        }
        
        async with httpx.AsyncClient() as client:
            try:
                response = await client.post(
                    f"{API_URLS['UPNIFY_BASE']}{ENDPOINTS['CREATE_PROSPECT']}",
                    headers={
                        "token": token,
                        "Content-Type": "application/json"
                    },
                    json=prospect_data
                )
                
                if not response.is_success:
                    error_text = response.text
                    raise Exception(f"Error al crear prospecto: {response.status_code} {response.reason_phrase}. {error_text}")
                
                result = response.json()
                return {
                    "success": True,
                    "message": "Prospecto creado exitosamente",
                    "data": result
                }
                
            except Exception as error:
                return {
                    "success": False,
                    "error": f"Error al crear prospecto en Upnify: {str(error)}"
                }
    
    async def search_contacts(self, search_term: str, limit: int = 10) -> Dict[str, Any]:
        """Search for contacts (prospects and clients)"""
        auth_info = await self.auth.get_token_and_user_info()
        token = auth_info["token"]
        
        if not search_term:
            raise ValueError("El término de búsqueda es obligatorio")
        
        if limit > DEFAULTS["MAX_SEARCH_RESULTS"]:
            limit = DEFAULTS["MAX_SEARCH_RESULTS"]
        
        async with httpx.AsyncClient() as client:
            try:
                response = await client.get(
                    f"{API_URLS['UPNIFY_BASE']}{ENDPOINTS['SEARCH_CONTACTS']}",
                    headers={
                        "token": token,
                        "Content-Type": "application/json"
                    },
                    params={
                        "buscar": search_term,
                        "cantidadRegistros": limit
                    }
                )
                
                if not response.is_success:
                    error_text = response.text
                    raise Exception(f"Error en búsqueda: {response.status_code} {response.reason_phrase}. {error_text}")
                
                contacts = response.json()
                
                return {
                    "success": True,
                    "message": f"Búsqueda completada. {len(contacts)} resultados encontrados",
                    "contacts": contacts,
                    "searchTerm": search_term,
                    "totalResults": len(contacts)
                }
                
            except Exception as error:
                return {
                    "success": False,
                    "error": f"Error al buscar contactos en Upnify: {str(error)}"
                }
